tests/threshold_sign.rs

```diff
@@ -13,30 +13,65 @@ use rand::{Rng, SeedableRng};
 
 type NodeId = u16;
 
-/// Tests a network of threshold signing instances with an optional expected value. Outputs the
-/// computed signature if the test is successful.
-fn test_threshold_sign<A>(
-    mut net: VirtualNet<ThresholdSign<NodeId>, A>,
-    mut rng: &mut TestRng,
-) -> Signature
+/// Creates a new test network.
+fn new_net<A, M, R>(
+    size: usize,
+    num_faulty: usize,
+    adv: A,
+    msg: M,
+    rng: &mut R,
+) -> VirtualNet<ThresholdSign<NodeId>, A>
 where
+    M: AsRef<[u8]> + 'static,
     A: Adversary<ThresholdSign<NodeId>>,
+    R: Rng,
 {
-    net.broadcast_input(&(), &mut rng)
-        .expect("threshold sign input failed");
+    NetBuilder::new(0..size as u16)
+        .num_faulty(num_faulty as usize)
+        .message_limit(size * (size - 1))
+        .no_time_limit()
+        .adversary(adv)
+        .using(move |node_info: NewNodeInfo<_>| {
+            ThresholdSign::new_with_document(Arc::new(node_info.netinfo), &msg)
+                .expect("Failed to create a ThresholdSign instance.")
+        })
+        .build(rng)
+        .expect("Could not construct test network.")
+        .0
+}
 
-    // Handle messages until all good nodes have terminated.
-    while !net.nodes().all(|node| node.algorithm().terminated()) {
-        let _ = net.crank_expect(&mut rng);
-    }
+/// Tests a network of threshold signing instances.
+fn test_threshold_sign_different_sizes<A, F>(new_adversary: F, seed: TestRngSeed)
+where
+    A: Adversary<ThresholdSign<NodeId>>,
+    F: Fn() -> A,
+{
+    // This returns an error in all but the first test.
+    let _ = env_logger::try_init();
 
-    // Verify that all instances output the same value.
-    let first = net.correct_nodes().nth(0).unwrap().outputs();
-    // TODO: Verify if signature is valid
-    assert!(!first.is_empty());
-    assert!(net.nodes().all(|node| node.outputs() == first));
+    let mut rng: TestRng = TestRng::from_seed(seed);
 
-    first[0].clone()
+    for &size in &[1, 2, 3, 4, 8, 15, 31] {
+        let num_faulty_nodes = util::max_faulty(size);
+        info!(
+            "Network size: {} good nodes, {} faulty nodes",
+            size - num_faulty_nodes,
+            num_faulty_nodes
+        );
+        let mut net = new_net(size, num_faulty_nodes, new_adversary(), "message", &mut rng);
+        net.broadcast_input(&(), &mut rng)
+            .expect("threshold sign input failed");
+        // Handle messages until all good nodes have terminated.
+        while !net.nodes().all(|node| node.algorithm().terminated()) {
+            let _ = net.crank_expect(&mut rng);
+        }
+        let node0 = net.correct_nodes().nth(0).unwrap();
+        // Verify that all instances output the same value.
+        assert_eq!(1, node0.outputs().len());
+        assert!(net.nodes().all(|node| node.outputs() == node0.outputs()));
+        let pk = node0.algorithm().netinfo().public_key_set().public_key();
+        assert!(pk.verify(&node0.outputs()[0], "message"));
```

Comment:
Maybe it's a minor detail but the expected value `"message"` could be cloned from the initial one to guarantee equality.